Change bulk's retry condition to be based on RestStatus

PnPie opened the pull request:
Previously bulk's retry logic was based on the exception type (`EsRejectedExecutionException`) of the failed response, this changes it to be based on RestStatus (`RestStatus.TOO_MANY_REQUESTS`), in order to support rest hight level client. (more information can be found #29254).

Close #28885

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

javanna commented on server/src/main/java/org/elasticsearch/action/bulk/Retry.java:
```diff
@@ -40,12 +40,12 @@
-    public Retry(Class<? extends Throwable> retryOnThrowable, BackoffPolicy backoffPolicy, Scheduler scheduler) {
-        this.retryOnThrowable = retryOnThrowable;
+    public Retry(RestStatus retryOnStatus, BackoffPolicy backoffPolicy, Scheduler scheduler) {
+        this.retryOnStatus = retryOnStatus;
```
shall we consider making this hardcoded rather than an argument given that we always pass in the same value for it?

PnPie replied:
Yes, as it's only used here in bulk for this case.

javanna commented on server/src/main/java/org/elasticsearch/action/bulk/Retry.java:
```diff
@@ -160,9 +160,8 @@ private boolean canRetry(BulkResponse bulkItemResponses) {
-                    final Throwable cause = bulkItemResponse.getFailure().getCause();
-                    final Throwable rootCause = ExceptionsHelper.unwrapCause(cause);
-                    if (!rootCause.getClass().equals(retryOnThrowable)) {
+                    final RestStatus status = bulkItemResponse.getFailure().getStatus();
```
just to double check: we don't need to unwrap here anymore because the status of the root cause is propagated to its ancestor?

PnPie replied:
Here we get the `status` field of `BulkItemResponse`'s `Failure`, it is a seperate field than `Exception cause` in `Failure` class. So I find it always has the good value (`RestStatus.TOO_MANY_REQUESTS`) ? because the exception type was changed only through `toXContent/fromXContent` of `BulkItemResponse`, but in it the `status` was already parsed seperatly. So the status should always be good ? (except after toXContent/fromXContent, the `BulkItemResponse` was transfered again using `readFrom/writeTo`, which I don't think it's the case ?)

If it's this, I changed `bulkItemResponse.getFailure().getStatus();` to `bulkItemResponse.status();`, because it's the same.

javanna replied:
I think this is correct. I also don't follow why `readFrom/writeTo` Would cause issues, the exception type does change but the status stays the same right?

javanna commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorRetryIT.java:
```diff
@@ -114,9 +114,9 @@ public void afterBulk(long executionId, BulkRequest request, Throwable failure)
-                        if (rootCause instanceof EsRejectedExecutionException) {
-                            if (rejectedExecutionExpected == false) {
+                        if (failure.getStatus() == RestStatus.TOO_MANY_REQUESTS) {
+                            if (!rejectedExecutionExpected) {
```
can you leave the previous `rejectedExecutionExpected == false` please? we prefer this one for readability.

javanna replied:
Could you address this please?

javanna commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorRetryIT.java:
```diff
@@ -114,9 +114,9 @@ public void afterBulk(long executionId, BulkRequest request, Throwable failure)
-                            if (rejectedExecutionExpected == false) {
+                        if (failure.getStatus() == RestStatus.TOO_MANY_REQUESTS) {
+                            if (!rejectedExecutionExpected) {
+                                Throwable rootCause = ExceptionsHelper.unwrapCause(failure.getCause());
```
maybe we do not even need to unwrap it here anymore? could we just use getCause instead when throwing assertion error below?

javanna replied:
I see that this has not been addressed. Is that on purpose? I see that you have done this on the client version of the test so it should be fine here too.

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/BulkProcessorRetryIT.java:
```diff
@@ -0,0 +1,221 @@
+
+    static {
+        System.setProperty("tests.rest.cluster", "localhost:9200");
+    }
```
this static block should go away. I don't see why it's needed.

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/BulkProcessorRetryIT.java:
```diff
@@ -0,0 +1,221 @@
+            assertThat(searchResultCount, lessThan(numberOfAsyncOps));
+        } else {
+            assertThat(searchResultCount, equalTo(numberOfAsyncOps));
+        }
```
the assertions seem more accurate here, thanks! Would you mind making the same change in the original test for the transport client? It should work there too right?

PnPie replied:
Yes !

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/BulkProcessorRetryIT.java:
```diff
@@ -0,0 +1,221 @@
+            assertThat(searchResultCount, equalTo(numberOfAsyncOps));
+        }
+
+    }
```
do you think that we should also have the last check based on the search API and the returned total hits? Or maybe now that we are using multi_get that step is not necessary?

PnPie replied:
Yeah, I think the `multi get request` we prepared when we indexed the documents is doing this, the same thing ? do a search (multi get) for all the indices in the end, to compare ? Seems in rest high level tests we are using rather multi get.

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/BulkProcessorRetryIT.java:
```diff
@@ -0,0 +1,221 @@
+        }
+
+        highLevelClient().indices().refresh(new RefreshRequest());
+        int searchResultCount = highLevelClient().multiGet(multiGetRequest).getResponses().length;
```
can you rename this variable, this is not about search anymore, rather the result of multi_get

javanna commented:
hi @PnPie would you have a chance to address my last comments? This is a good change, would love to get it in.

PnPie commented:
Hello @javanna,
Srry for the delay, in fact I had a very quick look last week-end but found a lot of strange things. I just re-check it today and I find I've had made a very ridiculous mistake :( I put the retry condition as `status == RETRY_STATUS -> fail` instead of `status != RETRY_STATUS -> fail` ...
I modified it and checked all the tests you mentioned passes, except `org.elasticsearch.client.BulkProcessorRetryIT` I have a TimeoutException, I feel like it might be my env problem ? So can we launch a test with jenkins to have a look ?

javanna commented:
test this please

javanna commented:
retest this please

javanna commented:
retest this please

javanna commented:
@PnPie tests were failing but not sure the failures were related to your changes, I merged master in and triggered a new build.

PnPie commented:
Hi @javanna,
Yes, seems sometimes `org.elasticsearch.action.bulk.BulkProcessorRetryIT` fails after adding the `rejectedAfterAllRetries` changes. I can look at it this week-end.

PnPie commented:
Hello @javanna,
I updated the PR and seems previously `BulkProcessorRetryIT` fails sometimes because we didn't take account of the "no retry (no back off)" situation. In this case when there is no retry for bulk request, the indexed docs could be equal to number of docs in bulk request (when there is no failures), or less than the number of docs in bulk request (when there is failures). I changed it as this.

javanna commented:
For the record, this requires #30384 otherwise the added tests fail.

javanna commented:
retest this please

javanna commented:
thanks a lot @PnPie !

jpountz commented:
@javanna This may be backported now.
